src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -170,23 +167,50 @@ internal abstract class AbstractPullDiagnosticHandler<TDiagnosticsParams, TRepor
             return CreateReturn(progress);
         }
 
-        private static Dictionary<Document, PreviousPullResult> GetDocumentToPreviousDiagnosticParams(
-            RequestContext context, ImmutableArray<PreviousPullResult> previousResults)
+        private static Dictionary<ProjectOrDocumentId, PreviousPullResult> GetDocumentToPreviousDiagnosticParams(
+            RequestContext context, ImmutableArray<PreviousPullResult> previousResults, out ImmutableArray<PreviousPullResult> removedDocuments)
         {
             Contract.ThrowIfNull(context.Solution);
 
-            var result = new Dictionary<Document, PreviousPullResult>();
+            var result = new Dictionary<ProjectOrDocumentId, PreviousPullResult>();
+            _ = ArrayBuilder<PreviousPullResult>.GetInstance(out var removedDocumentsBuilder);
             foreach (var diagnosticParams in previousResults)
             {
                 if (diagnosticParams.TextDocument != null)
                 {
-                    var document = context.Solution.GetDocument(diagnosticParams.TextDocument);
-                    if (document != null)
-                        result[document] = diagnosticParams;
+                    var id = GetIdForPreviousResult(diagnosticParams.TextDocument, context.Solution);
+                    if (id != null)
+                    {
+                        result[id.Value] = diagnosticParams;
+                    }
+                    else
+                    {
+                        // The client previously had a result from us for this document, but we no longer have it in our solution.
+                        // Record it so we can report to the client that it has been removed.
+                        removedDocumentsBuilder.Add(diagnosticParams);
+                    }
                 }
             }
 
+            removedDocuments = removedDocumentsBuilder.ToImmutable();
             return result;
+
+            static ProjectOrDocumentId? GetIdForPreviousResult(TextDocumentIdentifier textDocumentIdentifier, Solution solution)
```

Comment:
previous results can now be project files, so we have to check for both documents and projects when looking up the previous result identifier